.github/workflows/nightly.yml

```diff
@@ -0,0 +1,169 @@
+        with:
+          submodules: 'recursive'
+      - name: Set up Python 3.10
+        uses: actions/setup-python@v4
+        with:
+          python-version: "3.10"
+      - name: Setup virtual env
+        uses: ./.github/actions/make_init
+      - name: Run make develop
+        run: make develop
+      - name: Create tag
+        id: create_tag
+        run: |
+          git config --global user.email "core+streamlitbot-github@streamlit.io"
+          git config --global user.name "Streamlit Bot"
+
+          TAG="$(./scripts/pypi_nightly_create_tag.py)"
+          echo "::set-output name=tag::$TAG"
+
+          ./scripts/update_version.py $TAG
+          ./scripts/update_name.py streamlit-nightly
+
+          git add lib/setup.py
+          git add frontend/package.json
+
+          git add lib/streamlit/__init__.py
+          git add lib/streamlit/version.py
+
+          git commit -m "Update version and project name in files"
+
+          git tag -a $TAG -m "Streamlit nightly $TAG"
+          git push origin $TAG
+      # TODO: Using testing webhook, switch to actual st-open-source channel webhook later
+      - if: ${{ failure() }}
+        name: Nightly Tag Failure Slack Message
+        env: 
+          SLACK_WEBHOOK: ${{ secrets.TEST_SLACK_WEBHOOK }}
+        run: |
+          curl -X POST -H 'Content-type: application/json' --data '{"text":":blobonfire: Nightly job failed to create a tag"}' "$SLACK_WEBHOOK"
+
+  run-python-tests:
+    needs: create-nightly-tag
+    uses: ./.github/workflows/python-versions.yml
+    with:
+      ref: ${{needs.create-nightly-tag.outputs.tag}}
+
+  run-javascript-tests:
+    needs: create-nightly-tag
+    uses: ./.github/workflows/js-tests.yml
+    with:
+      ref: ${{needs.create-nightly-tag.outputs.tag}}
+
+  run-py-prod-deps-smoke-test:
+    needs: create-nightly-tag
+    uses: ./.github/workflows/py-prod-deps-smoke-test.yml
+    with:
+      ref: ${{needs.create-nightly-tag.outputs.tag}}
+
+  run-cypress-tests:
+    needs: create-nightly-tag
+    uses: ./.github/workflows/cypress.yml
+    with:
+      ref: ${{needs.create-nightly-tag.outputs.tag}}
+
+  test-status-notification:
+    runs-on: ubuntu-latest
+
+    if: ${{ always() }}
+    needs: [run-python-tests, run-javascript-tests, run-py-prod-deps-smoke-test, run-cypress-tests]
+
+    steps:
+      - name: Checkout Streamlit code
+        uses: actions/checkout@v3
+        with: 
+          persist-credentials: false
+          submodules: 'recursive'
+      - name: Send Slack notifications for failing tests
+        uses: ./.github/actions/slack
+        env:
+          SLACK_WEBHOOK: ${{ secrets.TEST_SLACK_WEBHOOK }}
+        with:
+          python_result: ${{ needs.run-python-tests.result }}
+          javascript_result: ${{ needs.run-javascript-tests.result }}
+          py_prod_deps_result: ${{ needs.run-py-prod-deps-smoke-test.result }}
+          cypress_result: ${{ needs.run-cypress-tests.result }}
+  
+  create-nightly-build:
```

Comment:
Do we want to avoid running this workflow if any of the tests fail? I think we'll currently try to publish the nightly regardless of test status, which we may or may not want to do.

Reply:
Agreed, we don't currently publish a nightly unless all the required tests pass & tag created successfully, and this workflow mirrors that logic (which I think makes sense). The `needs:` list allows us to identify any jobs that must complete successfully before the `create-nightly-build` job will run.

Reply:
Ah, okay. It's not knowing what the `if: ${{ always() }}` is for above that threw me off here. This is the default behavior I'd expect, then.